docs/IR.md

```diff
@@ -1,127 +1,372 @@
+
+Each operator set SHALL be defined in a separate document, also using protobuf as the serialization format. How operator set documents are found at runtime is implementation-dependent.
+
+__Note: As of the publication of this document, no ONNX implementation is known to process operator set documents.__
+
+The properties of an operator set are:
+
+Name|Type|Description
+|---|---|---|
+magic|string|The value ‘ONXXOPSET’
+ir_version|int32|The ONNX version corresponding to the operators.
+ir_version_prerelease|string|The prerelease component of the SemVer of the IR.
+ir_build_metadata|string|The symbolic identifier of the operator to invoke.
+domain|string|The domain of the operator set. Must be unique among all sets.
+opset_version|int64|The version of the set of operators. 
+doc_string|string|A human-readable documentation for this set of operators. Markdown is allowed.
+operator|Operator[]|The operators of this operator set.
+
+The operator set version is a simple integer value that is monotonically increased as new versions of the operator set are published.
+
+Operator sets other than the default operator set MUST specify its domain and use reverse domain names based on the responsible organization's identity, the same convention that is traditionally used for naming Java packages.
+
+### Operators
+
+Each operator used within a graph MUST be explicitly declared by one of the operator sets imported by the model.
+
+The properties of an operator definition are:
+
+Name|Type|Description
+|---|---|---|
+op_type|string|The name of the operator, as used in graph nodes. MUST be unique within the operator set’s domain.
+since_version|int64|The version of the operator set when this operator was introduced.
+status|OperatorStatus|One of ‘EXPERIMENTAL’ or ‘STABLE.’
+doc_string|string|A human-readable documentation string for this operator. Markdown is allowed.
+
+The version value MUST be the same value as the operator set version when the operator was first published. Subsequent versions of the operator set MUST NOT alter the signature or semantics of the operator once published as STABLE.
+
+The ‘status’ property indicates whether the syntax, semantics, or presence of the operator is in an experimental or stable stage. Once an operator is published as STABLE, it’s syntax and semantics MUST NOT change in subsequent versions of the operator set.
+
+There are two distinct ways to pass information to operators – inputs and attributes. The latter are used for values that are literal constants in the graph, while the former represent graph inputs or values computed elsewhere in the graph. This distinction may be highly relevant to achieving good performance for some implementations, while completely irrelevant to others.
```

Comment:
I think attributes should be non-trainable parameters? What do you think?

Reply:
Wei-Sheng,

I don't quite understand what you are proposing. Are you saying we shouldn't call them "attributes"? If so, I disagree -- we've been using that name since the beginning, and it has little to do with trainability.

Reply:
Sorry for not making it clear. My original opinion is to have a specific rule to distinguish "inputs" and "attributes" from each other. A "constant" is not ultimately clear to me because for example, LSTM's recursion matrices are constant under test mode but they are "inputs." Not sure if changing "literal constants" to "literal constants at inference stage and training stage" is helpful?

Reply:
Attributes are statically determined, and input can be dynamically computed. This is orthogonal to training and reference.

Reply:
Although calling a graph attribute constant literal sounds a little bit weird to me.

Reply:
Yes, I agree that "literal constants" is redundant, I'll fix that.
I don't understand your comment about 'training and reference,' where does the text mention that with regard to attributes or inputs?